discovery_engine/lib/src/domain/feed_manager.dart

```diff
@@ -145,18 +145,11 @@ class FeedManager {
       throw ArgumentError('source can\'t be empty');
     }
 
-    final allDocuments = await _docRepo.fetchAll();
-    final doesExist = allDocuments
-        .any((doc) => doc.isActive && doc.resource.sourceDomain == source);
-
-    if (!doesExist) {
-      throw ArgumentError('source $source not found in database');
-    }
-
```

Comment:
I removed this test as it's very prone to cause problems/complications for very little benefit.

It creates a implicit coupling with assumptions about the state a mostly unrelated part of the app will be in (the active documents).

This means any time we change any of the requirements around it it's prone to fail.

Furthermore it means in tests for add/rem excludedSources you have to recreate the right implicit state by happening to now how to combine the right fixtures and engine calls while avoiding filtering of any kind messing the state up. Which means there is a lot of implicit "state" in the test which makes it harder to maintain/understand.

(E.g. currently there is some unrelated problem somewhere, which makes creating that state hard/complicated.)

I.e. with this tests we can have false/spurious (but quite confusing) failures if:

- We allow adding sources in a different context (e.g. based on the /source endpoint of newscatcher).
- We change filtering of articles.
- We have a bug in filtering of sources, or the test server, or a unexpected change in the test fixture etc.
- We change how active documents are handled.

Reply:
The benefit is that we can get logic error from the user if they pass us invalid sources and there is not other way to know this otherwise.
The points you mentions are all valid but I don't agree with the solution. Instead of removing a check we should create better abstractions to allow to have simpler tests. We can introduce `SourcesRepository` to which we can ask if a source is ok or not. We can easily mock it in the tests and the real implementation will do what we are doing here and in the future we can change it to use the endpoint or some other data.

Reply:
I've mentioned a similar solution here https://github.com/xaynetwork/xayn_discovery_engine/pull/249#discussion_r828769317, which could be adopted to also take favorite sources into consideration. Maybe a smth like this?

```dart
enum SourceStatus {
  neutral, // default status
  excluded,
  favorite,
}

final sourcesRepo = {
  'example.com': SourceStatus.excluded,
  'bbc.co.uk': SourceStatus.favorite,
  'sport.de': SourceStatus.neutral,
};

final isValidSource = sourcesRepo['theguardian.com'] != null;
```

Reply:
I think we can refactor in this direction but using a repository instead of data structure. Maybe we don't need the neutral status depending on the api we want to expose but would be nice to have a central place for sources and where we can check that a source is only excluded or favorite and not both.

Reply:
Yes, ofc. I used data structure just as an example to describe how a Hive box could look like

Reply:
>  Instead of removing a check we should create better abstractions to allow to have simpler tests

Yes but not in this PR.

>  We can introduce SourcesRepository to which we can ask if a source is ok or not. We can easily mock it in the tests and the real implementation will do what we are doing here and in the future we can change it to use the endpoint or some other data.

Its only preventing a systematic API misuse by Team Blue which stems from sources being stringly typed, something we can change relatively easy without needing a add a bunch of complexity through another repository + mocking. Especially giving that mocking in integration tests across isolate boundaries just doesn't work well.